File: Demos/Visualizer/plotdata.cpp
#include "plotdata.h"
#include "visualizer.h"

#include <complex>
#include <iostream>

#include <QWidget>
#include <QVBoxLayout>
#include <QColor>
#include <QVector>

PlotData::PlotData(const std::size_t numChans, const int offset, 
                   QWidget *parent)
  : QWidget(parent)
{
    QWidget *window = new QWidget;
    setWindowTitle("Live Visualizer");
    
    QwtPlot *plot = new QwtPlot(parent);
      
    QVBoxLayout *layout = new QVBoxLayout;
    
    // colours for an octave
    QVector<QColor> colours = {QColor(255,0,0),
                               QColor(255,0,255),
                               QColor(255,136,0),
                               QColor(255,255,0),
                               QColor(85,255,0),
                               QColor(128,0,255),
                               QColor(0,153,0),
                               QColor(255,255,255),
                               QColor(0,178,178),
                               QColor(178,178,36),
                               QColor(0,170,255),
                               QColor(0,43,255)};
      
    // instead of seriesVector, store a vector of QVector<QPointF>
      
                                    
    // make all the required series and set their colours
    std::size_t c_idx; // colour index
    for (std::size_t i{0}; i<numChans; i++) {
        QwtPlotCurve *curve = new QwtPlotCurve();
        c_idx = (static_cast<int>(i)+offset) % 12;
        curve->setPen(colours[c_idx], 1.0);
        curves.append(curve);
        curve->attach(plot);
    }

    plot->setAxisScale(0, -50, 50);
    plot->setAxisScale(2, -50, 50);
    plot->updateAxes();
    
    showMaximized();
    layout->addWidget(plot);
    setLayout(layout);
    window->setLayout(layout);
    window->show();
}

void PlotData::update(const discriminator_t* frames,
                      const std::size_t numChans, 
                      const std::size_t numFrames)
{
    double x[numFrames];
    double y[numFrames];
    
    for (std::size_t i{0}; i<numChans; i++) {
        for (std::size_t n{0}; n<numFrames; n++){
            discriminator_t z {frames[(i*numFrames)+n]};
            x[n] = z.real();
            y[n] = z.imag();
        }
        curves[i]->setSamples(x, y, numFrames);
    }
//     plot->replot();
}

File: Demos/Visualizer/plotdata.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 25 15:48:23 2019

@author: keziah
"""

from PyQt5.QtWidgets import QDesktopWidget, QVBoxLayout, QWidget
import pyqtgraph as pg


class PlotData(QWidget):
    
    def __init__(self, numChans, offset):
        super().__init__()
        
        self.numChans = numChans
        self.offset = offset
        
        layout = QVBoxLayout()
        
#        p = pg.PlotWidget()
#        win.showMaximized()
        #win.resize(700,700)
        
        # Enable antialiasing for prettier plots
        pg.setConfigOptions(antialias=True)
        
        axr = 0.4 #1 #0.015
#        axrx = 1 #0.02
        
        self.c = [pg.hsvColor(0/360), 
                  pg.hsvColor(300/360),
                  pg.hsvColor(32/360), 
                  pg.hsvColor(60/360), 
                  pg.hsvColor(100/360), 
                  pg.hsvColor(270/360), 
                  pg.hsvColor(120/360, val=0.6), 
                  pg.hsvColor(160/360, sat=0.0), 
                  pg.hsvColor(180/360, val=0.7), 
                  pg.hsvColor(60/360, sat=0.8, val=0.7),
                  pg.hsvColor(200/360), 
                  pg.hsvColor(230/360)]
        
        self.plotWidget = pg.PlotWidget()
        for i in range(self.numChans):
            colour = self.c[(i+self.offset)%12]
            self.plotWidget.plot(pen=colour)
            
        self.plotWidget.setRange(xRange=(-axr, axr), yRange=(-axr, axr))
            
        plotItem = self.plotWidget.getPlotItem()
        self.dataItems = plotItem.listDataItems()
        
        layout.addWidget(self.plotWidget)
        self.setLayout(layout)
        self.show()
        
        

    def centre(self):
        """ Centre window on screen. """
        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())
        
        
    def update(self, z):
        
        for k in range(self.numChans):
#            self.dataItems[k].setData(z[k])
            
            x = z[k].real[::2]
            y = z[k].imag[::2]
            self.dataItems[k].setData(x, y)
            
